File: Ios/BarkCloud/BarkCloud/Features/Files/SmartFolders/SmartFolderDetailViewModel.swift
import Foundation
import Observation
import BarkCloudKit

/// Содержимое умной папки: файлы по критериям с cursor-пагинацией. Рендер (сетка
/// или список) определяется `viewMode` папки в экране.
@MainActor
@Observable
final class SmartFolderDetailViewModel {
    var items: [MediaItem] = []
    var isLoading = true
    var isLoadingMore = false
    var snackbar: String?

    private var canLoadMore = false
    private var cursorCreatedAt: Date?
    private var cursorFileID = ""

    private let folderID: String
    private let repo: DynamicFolderRepository
    private var didLoad = false

    init(folderID: String, repo: DynamicFolderRepository) {
        self.folderID = folderID
        self.repo = repo
    }

    func loadIfNeeded() async {
        guard !didLoad else { return }
        didLoad = true
        await reload()
    }

    func reload(showSpinner: Bool = true) async {
        if showSpinner { isLoading = true }
        do {
            let page = try await repo.listItems(folderID: folderID, limit: 60)
            items = page.items.map(MediaItem.init(asset:))
            cursorCreatedAt = page.nextCursorCreatedAt
            cursorFileID = page.nextCursorFileID
            canLoadMore = page.hasMore
        } catch {
            items = []
            snackbar = domainErrorMessage(error)
        }
        isLoading = false
    }

    func loadMoreIfNeeded(current item: MediaItem) async {
        guard canLoadMore, !isLoadingMore, item.id == items.last?.id else { return }
        isLoadingMore = true
        do {
            let page = try await repo.listItems(
                folderID: folderID, limit: 60,
                cursorCreatedAt: cursorCreatedAt, cursorFileID: cursorFileID
            )
            items.append(contentsOf: page.items.map(MediaItem.init(asset:)))
            cursorCreatedAt = page.nextCursorCreatedAt
            cursorFileID = page.nextCursorFileID
            canLoadMore = page.hasMore
        } catch {
            snackbar = domainErrorMessage(error)
        }
        isLoadingMore = false
    }

    func snackbarShown() { snackbar = nil }
}

File: Android/BarkCloud.Android/app/src/main/java/com/barkfluff/BarkCloud/ui/smartfolders/SmartFolderDetailViewModel.kt
package com.barkfluff.BarkCloud.ui.smartfolders

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.CreationExtras
import com.barkfluff.BarkCloud.BarkCloudApplication
import com.barkfluff.BarkCloud.data.cloud.DynamicFolderRepository
import com.barkfluff.BarkCloud.data.cloud.MediaAsset
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class SmartFolderDetailUiState(
    val isLoading: Boolean = true,
    val isRefreshing: Boolean = false,
    val isLoadingMore: Boolean = false,
    val canLoadMore: Boolean = true,
    val items: List<MediaAsset> = emptyList(),
    val snackbar: String? = null,
) {
    val isEmpty: Boolean get() = !isLoading && items.isEmpty()
}

class SmartFolderDetailViewModel(
    private val repository: DynamicFolderRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(SmartFolderDetailUiState())
    val state: StateFlow<SmartFolderDetailUiState> = _state.asStateFlow()

    private var folderId = ""
    private var started = false
    private var cursorCreatedAt: Long? = null
    private var cursorFileId = ""

    fun start(folderId: String) {
        if (started) return
        started = true
        this.folderId = folderId
        reload()
    }

    fun reload() {
        cursorCreatedAt = null
        cursorFileId = ""
        viewModelScope.launch {
            _state.update { it.copy(isLoading = it.items.isEmpty(), isRefreshing = it.items.isNotEmpty()) }
            runCatching { repository.listItems(folderId) }
                .onSuccess { page ->
                    cursorCreatedAt = page.nextCursorCreatedAtMillis
                    cursorFileId = page.nextCursorFileId
                    _state.update {
                        it.copy(
                            isLoading = false,
                            isRefreshing = false,
                            items = page.items,
                            canLoadMore = page.hasMore,
                        )
                    }
                }
                .onFailure { error ->
                    _state.update {
                        it.copy(isLoading = false, isRefreshing = false, snackbar = error.message)
                    }
                }
        }
    }

    fun loadMore() {
        val s = _state.value
        if (!s.canLoadMore || s.isLoading || s.isLoadingMore) return
        _state.update { it.copy(isLoadingMore = true) }
        viewModelScope.launch {
            runCatching {
                repository.listItems(
                    folderId = folderId,
                    cursorCreatedAtMillis = cursorCreatedAt,
                    cursorFileId = cursorFileId,
                )
            }
                .onSuccess { page ->
                    cursorCreatedAt = page.nextCursorCreatedAtMillis
                    cursorFileId = page.nextCursorFileId
                    _state.update {
                        it.copy(
                            isLoadingMore = false,
                            items = it.items + page.items,
                            canLoadMore = page.hasMore,
                        )
                    }
                }
                .onFailure { error ->
                    _state.update { it.copy(isLoadingMore = false, snackbar = error.message) }
                }
        }
    }

    fun snackbarShown() = _state.update { it.copy(snackbar = null) }

    companion object {
        fun factory(): ViewModelProvider.Factory = object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>, extras: CreationExtras): T {
                val app = extras[ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY]
                    as BarkCloudApplication
                return SmartFolderDetailViewModel(app.dynamicFolderRepository) as T
            }
        }
    }
}
